pkg/util/pki/asn1_util_test.go

```diff
@@ -0,0 +1,240 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pki
+
+import (
+	"encoding/asn1"
+	"errors"
+	"reflect"
+	"testing"
+)
+
+func TestParseObjectIdentifier(t *testing.T) {
+	testCases := []struct {
+		oidString   string
+		expectedOid asn1.ObjectIdentifier
+		expectedErr error
+	}{
+		{
+			oidString:   "1.2.3.4.5",
+			expectedOid: asn1.ObjectIdentifier{1, 2, 3, 4, 5},
+			expectedErr: nil,
+		},
+		{
+			oidString:   "1.2.840.113549.1.1.1",
+			expectedOid: asn1.ObjectIdentifier{1, 2, 840, 113549, 1, 1, 1},
+			expectedErr: nil,
+		},
+		{
+			oidString:   "1.3.6.1.4.1.311.60.2.1.3",
+			expectedOid: asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 311, 60, 2, 1, 3},
+			expectedErr: nil,
+		},
+		{
+			oidString:   ".",
+			expectedOid: nil,
+			expectedErr: errors.New("strconv.Atoi: parsing \"\": invalid syntax"),
+		},
+		{
+			oidString:   ".555",
+			expectedOid: nil,
+			expectedErr: errors.New("strconv.Atoi: parsing \"\": invalid syntax"),
+		},
+		{
+			oidString:   "555.",
+			expectedOid: nil,
+			expectedErr: errors.New("strconv.Atoi: parsing \"\": invalid syntax"),
+		},
+		{
+			oidString:   "test.5",
+			expectedOid: nil,
+			expectedErr: errors.New("strconv.Atoi: parsing \"test\": invalid syntax"),
+		},
+	}
+
+	for _, tc := range testCases {
+		oid, err := ParseObjectIdentifier(tc.oidString)
+		if err != nil {
+			if tc.expectedErr == nil {
+				t.Errorf("Unexpected error: %v", err)
+			} else if err.Error() != tc.expectedErr.Error() {
+				t.Errorf("Expected error: %v, got: %v", tc.expectedErr, err)
+			}
+		} else if !oid.Equal(tc.expectedOid) {
+			t.Errorf("Expected OID: %v, got: %v", tc.expectedOid, oid)
+		}
+	}
+}
+
+func TestMarshalAndUnmarshalUniversalValue(t *testing.T) {
+	testCases := []struct {
+		name                string
+		uv                  UniversalValue
+		raw                 asn1.RawValue
+		overrideRoundtripUv *UniversalValue
+	}{
+		{
+			name: "Test with IA5String",
+			uv: UniversalValue{
+				Ia5String: "test",
+			},
+			raw: asn1.RawValue{
+				Bytes: []byte("test"),
+				Class: asn1.ClassUniversal,
+				Tag:   asn1.TagIA5String,
+			},
+		},
+		{
+			name: "Test with Utf8String",
+			uv: UniversalValue{
+				Utf8String: "test",
+			},
+			raw: asn1.RawValue{
+				Bytes: []byte("test"),
+				Class: asn1.ClassUniversal,
+				Tag:   asn1.TagUTF8String,
+			},
+		},
+		{
+			name: "Test with PrintableString",
+			uv: UniversalValue{
+				PrintableString: "test",
+			},
+			raw: asn1.RawValue{
+				Bytes: []byte("test"),
+				Class: asn1.ClassUniversal,
+				Tag:   asn1.TagPrintableString,
+			},
+		},
+		{
+			name: "Test with Bytes",
+			uv: UniversalValue{
+				Bytes: []byte{0x16, 0x04, 0x74, 0x65, 0x73, 0x74},
+			},
```

Comment:
suggestion: I think this byte slice would benefit from a comment explaining what it is. It looks to me like an ASN.1 prefix followed by the word "test" - is that right?

Reply:
@inteon i think you have more context on this one